### PR TITLE
feat: add sdk infomation metrics

### PR DESCRIPTION
**Pull request by whynowy**
For MonoVertex:
```
# HELP sdk_info A metric with a constant value '1', labeled by SDK information such as version, language, and type.
# TYPE sdk_info gauge
sdk_info{component="mono-vertex",component_name="simple-mono-vertex",language="rust",version="0.1.1",type="sourcer"} 1
sdk_info{component="mono-vertex",component_name="simple-mono-vertex",language="rust",version="0.1.1",type="sinker"} 1
sdk_info{component="mono-vertex",component_name="simple-mono-vertex",language="rust",version="0.1.1",type="sourcetransformer"} 1
```

For Vertex:
```
# HELP sdk_info A metric with a constant value '1', labeled by SDK information such as version, language, and type
# TYPE sdk_info gauge
sdk_info{component="vertex",component_name="simple-pipeline-cat",language="go",type="mapper",version="v0.9.0"} 1
```

Similarly, there's a metric emitted for side inputs manager.

### CODE REVIEW

File: rust/numaflow-core/src/shared/server_info.rs [hunk 1]
```diff
@@ -18,6 +18,17 @@ use crate::shared::server_info::version::SdkConstraints;
 // Equivalent to U+005C__END__.
 const END: &str = "U+005C__END__";
 
+pub const CONTAINER_TYPE_SOURCER: &str = "sourcer";
```
Comment (whynowy) [inline rust/numaflow-core/src/shared/server_info.rs:21]: What is the best way to do this in rust?

Comment (kohlisid) [reply rust/numaflow-core/src/shared/server_info.rs:21]: Enum?

Comment (whynowy) [reply rust/numaflow-core/src/shared/server_info.rs:21]: Enum with string? Not seems to work.

Comment (kohlisid) [reply rust/numaflow-core/src/shared/server_info.rs:21]: Do you want something like this?
https://github.com/numaproj/numaflow/pull/2020

**Comment by KeranYang** [comment]
do we need extra pipeline and vertex name dimensions? one pipeline can have multiple vertices using different SDK versions

**Comment by KeranYang** [comment]
And just curious, what is the main use case of introducing this metric?

File: rust/numaflow-core/src/shared/server_info.rs (hunk 1 above)
Comment (whynowy) [reply rust/numaflow-core/src/shared/server_info.rs:21]: Constant seems to be simpler for this specific use case.

**Comment by whynowy** [comment]
> do we need extra pipeline and vertex name dimensions? one pipeline can have multiple vertices using different SDK versions

No need for that I think.

File: rust/numaflow-core/src/metrics.rs [hunk 2]
```diff
@@ -366,6 +400,21 @@ pub(crate) fn forward_pipeline_metrics() -> &'static PipelineMetrics {
     PIPELINE_METRICS.get_or_init(PipelineMetrics::new)
 }
 
+
+// forward_metrics_labels is a helper function used to fetch the
+// SDK_INFO_LABELS object
```
Comment (vigith) [inline rust/numaflow-core/src/metrics.rs:404]: wrong comment?

File: rust/numaflow-core/src/pipeline.rs [hunk 3]
```diff
@@ -267,11 +270,10 @@ async fn create_transformer(
         if let config::components::transformer::TransformerType::UserDefined(ud_transformer) =
             &transformer_config.transformer_type
         {
-            check_for_server_compatibility(
-                ud_transformer.socket_path.clone().into(),
-                cln_token.clone(),
-            )
-            .await?;
+            _ = sdk_server_info(ud_transformer.socket_path.clone().into(), cln_token.clone())
+                .await?;
```
Comment (vigith) [inline rust/numaflow-core/src/pipeline.rs:274]:
```suggestion
            _ = sdk_server_info(ud_transformer.socket_path.clone().into(), cln_token.clone())
                .await?;
```

do we need this change?

File: rust/numaflow-core/src/shared/server_info.rs (hunk 1 above)
Comment (vigith) [reply rust/numaflow-core/src/shared/server_info.rs:21]: idiomatic rust is to use enum and `impl std::fmt::Display` for getting the string version string.

File: rust/numaflow-core/src/shared/utils.rs [hunk 4]
```diff
@@ -186,16 +186,15 @@ pub(crate) async fn create_sink_handle(
             None,
         )),
         SinkType::UserDefined(ud_config) => {
-            check_for_server_compatibility(
-                ud_config.server_info_path.clone().into(),
-                cln_token.clone(),
-            )
-            .await?;
+            _ = sdk_server_info(ud_config.server_info_path.clone().into(), cln_token.clone())
+                .await?;
```
Comment (vigith) [inline rust/numaflow-core/src/shared/utils.rs:190]: any reason for this change?

File: rust/numaflow-core/src/pipeline.rs (hunk 3 above)
Comment (whynowy) [reply rust/numaflow-core/src/pipeline.rs:274]: Do we need to handle the the return in rust? Also, I assume something needs to be added to for the returned server info later on - this is for pipeline.

File: rust/numaflow-core/src/shared/utils.rs (hunk 4 above)
Comment (whynowy) [reply rust/numaflow-core/src/shared/utils.rs:190]: Same reason as the other one.

File: rust/numaflow-core/src/pipeline.rs (hunk 3 above)
Comment (vigith) [reply rust/numaflow-core/src/pipeline.rs:274]: `?` does the return on error. the success is a zero-size-type `()` which has no meaning.

Comment (whynowy) [reply rust/numaflow-core/src/pipeline.rs:274]: The success is changed to ServerInfo.

Comment (vigith) [reply rust/numaflow-core/src/pipeline.rs:274]: then it is fine, i missed that

**Review by vigith — APPROVED** [review]

**Pull request merged**